### PR TITLE
HPCC-22148 Add ActivityContext support for embedded Java code

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Added new tests

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

File: testing/regress/ecl/java-activity.ecl [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+//class=embedded
+//class=3rdparty
+//nothor     - behaviour depends on number of slaves
```
Comment (ghalliday) [inline testing/regress/ecl/java-activity.ecl]: This should really support thor -  by checking the activity parameters to see if it is not node 0, and not generate any data.

File: testing/regress/ecl/java-activity.ecl [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+//noroxie
```
Comment (ghalliday) [inline testing/regress/ecl/java-activity.ecl]: Why not supported in roxie?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman looks good.  It should be quite easy to get the example working with thor.

File: testing/regress/ecl/java-activity.ecl (hunk 2 above)
Comment (richardkchapman) [reply testing/regress/ecl/java-activity.ecl]: It should work in Roxie. I can't remember why I disabled it

**Comment by richardkchapman** [comment]
@ghalliday Improved the test (and stumbled on a couple of bugs in the process)

File: ecl/hqlcpp/hqlcpp.cpp [hunk 3]
```diff
@@ -12068,8 +12068,9 @@ void HqlCppTranslator::buildScriptFunctionDefinition(BuildCtx &ctx, IHqlExpressi
     else
         embedOptions.setown(getEmbedOptionString(bodyCode));
     IValue *optionVal = embedOptions->queryValue();
+    bool isActivity = functionBodyIsActivity(bodyCode);
 
-    bool threadlocal = bodyCode->hasAttribute(_threadlocal_Atom) && queryVal != nullptr && optionVal != nullptr;
+    bool threadlocal = bodyCode->hasAttribute(_threadlocal_Atom) && queryVal != nullptr && optionVal != nullptr && !isActivity;
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlcpp.cpp]: This isn't necessarily the perfect fix - we could pass in the activity separately instead, and then still keep the threadlocal. But it's probably good enough and keeps things compatible. When using activity flag the chances are it's not being reused a lot anyway.

**Comment by ghalliday** [comment]
@richardkchapman please squash

**Comment by richardkchapman** [comment]
@ghalliday Squashed

**Pull request merged**